osu.Game/Database/UserLookupCache.cs

```diff
@@ -47,70 +40,109 @@ private Task<List<User>> getQueryTaskForUser(int userId)
         {
             lock (taskAssignmentLock)
             {
-                nextTaskIDs.Add(userId);
+                // attempt to queue on the next pending task.
+                var lastTask = tasks.LastOrDefault();
 
-                // if there's a pending request which hasn't been started yet (and is not yet full), we can wait on it.
-                if (pendingRequest != null && !pendingRequestConsumedIDs && nextTaskIDs.Count < 50)
-                    return pendingRequest;
+                if (lastTask?.AddUser(userId) == true)
+                    return lastTask.Task;
 
-                return queueNextTask(nextLookup);
-            }
+                var lookup = new LookupTask(api);
 
-            List<User> nextLookup()
-            {
-                int[] lookupItems;
-
-                lock (taskAssignmentLock)
+                // always start the next task running when a previous task finishes.
+                lookup.Task.ContinueWith(completed =>
                 {
-                    pendingRequestConsumedIDs = true;
-                    lookupItems = nextTaskIDs.ToArray();
-                    nextTaskIDs.Clear();
-
-                    if (lookupItems.Length == 0)
-                    {
-                        queueNextTask(null);
-                        return new List<User>();
-                    }
-                }
+                    var dequeued = tasks.Dequeue();
```

Comment:
This code probably needs to be locked? This continuation is going to run on the global threadpool is it not?

Reply:
oh good point, i didn't even consider the queue's thread safety here (although i believe it won't fail on a dequeue since it's guaranteed to have an item).

do you prefer locking or `ConcurrentQueue`?

Reply:
Are you still going to have to lock the rest of the code? If so then I'd prefer locking to keep it simple and one-pattern. Probably extract the callback into a separate method while you're at it.